keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -930,31 +960,40 @@ Not that this matters much, because dispatching for those is not
 actually limited.  The sum of those limits, however, is subtracted
 from `ServerCL` to produce a value called `RemainingServerCL` that is
 used in computing the allocations for the non-exempt priority levels.
-If `RemainingServerCL` is zero or negative then all the non-exempt
-priority levels get `CurrentCL = 0`.  Otherwise, the computation
-proceeds as follows.
 
 Because of the borrowing by exempt priority levels, lower bounds could
-be problematic.  Define `LowerBoundSum` as follows.
+be problematic.  Define `MinCLSum` and `MinCurrentCLSum` as follows.
 
 ```
-LowerBoundSum = sum[non-exempt priority level i] MinCurrentCL(i)
+MinCLSum        = sum[non-exempt priority level i] MinCL(i)
+MinCurrentCLSum = sum[non-exempt priority level i] MinCurrentCL(i)
 ```
 
-If `LowerBoundSum = RemainingServerCL` then there is no wiggle room:
-each non-exempt priority level gets `CurrentCL = MinCurrentCL`.
+When `RemainingServerCL < MinCLSum`, the expected usage by the exempt
+levels is so high that we would like to give the non-exempt levels
+less than their minimum. But that could be problematic. So in this
+case and when `RemainingServerCL = MinCLSum` we set `CurrentCL` for
+the non-exempt levels by the following formula.
+
+```
+CurrentCL(i) = MinCL(i)
+```
 
-If `LowerBoundSum > RemainingServerCL` then the problem is
-over-constrained.  The solution taken is to reduce all the lower
-bounds in the same proportion, to the point where their sum is
-feasible.  At that point, there is no wiggle room.  Thus, in this case
-the settings are as follows.
+When `MinCLSum < RemainingServerCL < MinCurrentCLSum`, the concurrency
+available to non-exempt levels is above their rock-bottom limit but
+not enough to give each its `MinCurrentCL`. In this case and when
+`RemainingServerCL = MinCurrentCLSum` we share the excess above
+`MinCL` proportionally to each level's deficit and weight, using the
+following formulae.
 
 ```
-CurrentCL(i) = MinCurrentCL(i) * RemainingServerCL / LowerBoundSum
+CurrentCL(i) = MinCL(i) + ( MinCurrentCL(i) - MinCL(i) ) * Weight(i) *
+                          (RemainingServerCL-MinCLSum) / WeightedDiffSum
+WeightedDiffSum = sum[non-exempt priority level i]
+                          ( MinCurrentCL(i) - MinCL(i) ) * Weight(i)
```

Comment:
I might not have fully understood those formulae, but questions:

* why can't we just apply the previous logic -- scaling down the CurrentCL proportional to `RemainingServerCL / LowerBoundSum`? why do the weights have to impact this if no borrowing is happening here.
* A more direct question: I tested APF a while ago and learned a rule of "the owner keeps the shares if inuse", now would the weights break this rule? i.e. PLs with lower weights will be borrowed even when their demands are higher than their NominalCL?

Reply:
First, remember that the PR that you are looking at is actually combining two changes: one is #4623 , which aims to heal the long-standing divergence between code and KEP by moving both to a middle ground, and then there is the change to add weighting to the priority levels.

It is true that when `MinCLSum < RemainingServerCL < MinCurrentCLSum` there is no borrowing between non-exempt priority levels, but the exempt level(s) have borrowed so much from the non-exempt ones that each must give up some concurrency. So it is actually a borrowing scenario.

Remember that you have not tested the behavior described in #4623 unless you tested its implementation in https://github.com/kubernetes/kubernetes/pull/124736 .

Regarding "the owner keeps the shares if in use", I would be a little more precise: the rule is that each non-exempt level gets its _nominal_ concurrency limit if at least that much has been in use AND the borrowing by exempt levels is not so great as to cut into that. The formulae that you have commented. on concern the case where the exempt level(s) have borrowed so much that the non-exempt ones can not get their nominal concurrency limits.